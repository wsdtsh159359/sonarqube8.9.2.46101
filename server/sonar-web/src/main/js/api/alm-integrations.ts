/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { get, getJSON, parseError, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  AzureProject,
  AzureRepository,
  BitbucketProject,
  BitbucketRepository,
  GithubOrganization,
  GithubRepository,
  GitlabProject
} from '../types/alm-integration';
import { ProjectBase } from './components';

export function setAlmPersonalAccessToken(almSetting: string, pat: string): Promise<void> {
  return post('/api/alm_integrations/set_pat', { almSetting, pat }).catch(throwGlobalError);
}

export function checkPersonalAccessTokenIsValid(
  almSetting: string
): Promise<{ status: boolean; error?: string }> {
  return get('/api/alm_integrations/check_pat', { almSetting })
    .then(() => ({ status: true }))
    .catch(async (response: Response) => {
      if (response.status === 400) {
        const error = await parseError(response);
        return { status: false, error };
      }
      return throwGlobalError(response);
    });
}

export function getAzureProjects(almSetting: string): Promise<{ projects: AzureProject[] }> {
  return getJSON('/api/alm_integrations/list_azure_projects', { almSetting }).catch(
    throwGlobalError
  );
}

export function getAzureRepositories(
  almSetting: string,
  projectName: string
): Promise<{ repositories: AzureRepository[] }> {
  return getJSON('/api/alm_integrations/search_azure_repos', { almSetting, projectName }).catch(
    throwGlobalError
  );
}

export function searchAzureRepositories(
  almSetting: string,
  searchQuery: string
): Promise<{ repositories: AzureRepository[] }> {
  return getJSON('/api/alm_integrations/search_azure_repos', { almSetting, searchQuery }).catch(
    throwGlobalError
  );
}

export function importAzureRepository(
  almSetting: string,
  projectName: string,
  repositoryName: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_azure_project', {
    almSetting,
    projectName,
    repositoryName
  }).catch(throwGlobalError);
}

export function getBitbucketServerProjects(
  almSetting: string
): Promise<{ projects: BitbucketProject[] }> {
  return getJSON('/api/alm_integrations/list_bitbucketserver_projects', { almSetting });
}

export function getBitbucketServerRepositories(
  almSetting: string,
  projectName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    projectName
  });
}

export function importBitbucketServerProject(
  almSetting: string,
  projectKey: string,
  repositorySlug: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_bitbucketserver_project', {
    almSetting,
    projectKey,
    repositorySlug
  }).catch(throwGlobalError);
}

export function searchForBitbucketServerRepositories(
  almSetting: string,
  repositoryName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    repositoryName
  });
}

export function getGithubClientId(almSetting: string): Promise<{ clientId?: string }> {
  return getJSON('/api/alm_integrations/get_github_client_id', { almSetting });
}

export function importGithubRepository(
  almSetting: string,
  organization: string,
  repositoryKey: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_github_project', {
    almSetting,
    organization,
    repositoryKey
  }).catch(throwGlobalError);
}

export function getGithubOrganizations(
  almSetting: string,
  token: string
): Promise<{ organizations: GithubOrganization[] }> {
  return getJSON('/api/alm_integrations/list_github_organizations', {
    almSetting,
    token
  }).catch((response?: Response) => {
    if (response && response.status !== 400) {
      throwGlobalError(response);
    }
  });
}

export function getGithubRepositories(data: {
  almSetting: string;
  organization: string;
  pageSize: number;
  page?: number;
  query?: string;
}): Promise<{ repositories: GithubRepository[]; paging: T.Paging }> {
  const { almSetting, organization, pageSize, page = 1, query } = data;
  return getJSON('/api/alm_integrations/list_github_repositories', {
    almSetting,
    organization,
    p: page,
    ps: pageSize,
    q: query || undefined
  }).catch(throwGlobalError);
}

export function getGitlabProjects(data: {
  almSetting: string;
  page?: number;
  pageSize?: number;
  query?: string;
}): Promise<{ projects: GitlabProject[]; projectsPaging: T.Paging }> {
  const { almSetting, pageSize, page, query } = data;
  return getJSON('/api/alm_integrations/search_gitlab_repos', {
    almSetting,
    projectName: query || undefined,
    p: page,
    ps: pageSize
  })
    .then(({ repositories, paging }) => ({ projects: repositories, projectsPaging: paging }))
    .catch(throwGlobalError);
}

export function importGitlabProject(data: {
  almSetting: string;
  gitlabProjectId: string;
}): Promise<{ project: ProjectBase }> {
  const { almSetting, gitlabProjectId } = data;
  return postJSON('/api/alm_integrations/import_gitlab_project', {
    almSetting,
    gitlabProjectId
  }).catch(throwGlobalError);
}
