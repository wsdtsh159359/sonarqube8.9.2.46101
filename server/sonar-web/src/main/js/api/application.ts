/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { Application, ApplicationPeriod, ApplicationProject } from '../types/application';
import { Visibility } from '../types/component';

export function getApplicationLeak(
  application: string,
  branch?: string
): Promise<ApplicationPeriod[]> {
  return getJSON('/api/applications/show_leak', { application, branch }).then(
    r => r.leaks,
    throwGlobalError
  );
}

export function getApplicationDetails(application: string, branch?: string): Promise<Application> {
  return getJSON('/api/applications/show', { application, branch }).then(
    r => r.application,
    throwGlobalError
  );
}

export function addApplicationBranch(data: {
  application: string;
  branch: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/create_branch', data).catch(throwGlobalError);
}

export function updateApplicationBranch(data: {
  application: string;
  branch: string;
  name: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/update_branch', data).catch(throwGlobalError);
}

export function deleteApplicationBranch(application: string, branch: string) {
  return post('/api/applications/delete_branch', { application, branch }).catch(throwGlobalError);
}

export function getApplicationProjects(data: {
  application: string;
  p?: number;
  ps?: number;
  q?: string;
  selected: string;
}): Promise<{ paging: T.Paging; projects: ApplicationProject[] }> {
  return getJSON('/api/applications/search_projects', data).catch(throwGlobalError);
}

export function addProjectToApplication(application: string, project: string) {
  return post('/api/applications/add_project', { application, project }).catch(throwGlobalError);
}

export function removeProjectFromApplication(application: string, project: string) {
  return post('/api/applications/remove_project', { application, project }).catch(throwGlobalError);
}

export function refreshApplication(key: string) {
  return post('/api/applications/refresh', { key }).catch(throwGlobalError);
}

export function createApplication(
  name: string,
  description: string,
  key: string | undefined,
  visibility: string
): Promise<{
  application: {
    description?: string;
    key: string;
    name: string;
    visibility: Visibility;
  };
}> {
  return postJSON('/api/applications/create', { description, key, name, visibility }).catch(
    throwGlobalError
  );
}

export function deleteApplication(application: string) {
  return post('/api/applications/delete', { application }).catch(throwGlobalError);
}

export function editApplication(application: string, name: string, description: string) {
  return post('/api/applications/update', { name, description, application }).catch(
    throwGlobalError
  );
}
