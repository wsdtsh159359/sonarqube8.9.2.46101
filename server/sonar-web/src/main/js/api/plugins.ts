/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { findLastIndex } from 'lodash';
import { getJSON, post } from 'sonar-ui-common/helpers/request';
import { isDefined } from 'sonar-ui-common/helpers/types';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  AvailablePlugin,
  InstalledPlugin,
  PendingPluginResult,
  PluginType,
  Update
} from '../types/plugins';

export function getAvailablePlugins(): Promise<{
  plugins: AvailablePlugin[];
  updateCenterRefresh: string;
}> {
  return getJSON('/api/plugins/available').catch(throwGlobalError);
}

export function getPendingPlugins(): Promise<PendingPluginResult> {
  return getJSON('/api/plugins/pending').catch(throwGlobalError);
}

function getLastUpdates(updates: undefined | Update[]): Update[] {
  if (!updates) {
    return [];
  }
  const lastUpdate = ['COMPATIBLE', 'REQUIRES_SYSTEM_UPGRADE', 'DEPS_REQUIRE_SYSTEM_UPGRADE'].map(
    status => {
      const index = findLastIndex(updates, update => update.status === status);
      return index > -1 ? updates[index] : undefined;
    }
  );
  return lastUpdate.filter(isDefined);
}

function addChangelog(update: Update, updates?: Update[]) {
  if (!updates) {
    return update;
  }
  const index = updates.indexOf(update);
  const previousUpdates = index > 0 ? updates.slice(0, index) : [];
  return { ...update, previousUpdates };
}

function getInstalledPluginApi(type = PluginType.External) {
  return getJSON('/api/plugins/installed', { f: 'category', type });
}

function getUpdatesPluginApi() {
  return getJSON('/api/plugins/updates');
}

export function getInstalledPlugins(
  type: PluginType = PluginType.External
): Promise<InstalledPlugin[]> {
  return getInstalledPluginApi(type).then(({ plugins }) => plugins, throwGlobalError);
}

export function getInstalledPluginsWithUpdates(): Promise<InstalledPlugin[]> {
  return Promise.all([getInstalledPluginApi(), getUpdatesPluginApi()])
    .then(([installed, updates]) =>
      installed.plugins.map((plugin: InstalledPlugin) => {
        const updatePlugin: InstalledPlugin = updates.plugins.find(
          (p: InstalledPlugin) => p.key === plugin.key
        );
        if (updatePlugin) {
          return {
            ...updatePlugin,
            ...plugin,
            updates: getLastUpdates(updatePlugin.updates).map(update =>
              addChangelog(update, updatePlugin.updates)
            )
          };
        }
        return plugin;
      })
    )
    .catch(throwGlobalError);
}

export function getPluginUpdates(): Promise<InstalledPlugin[]> {
  return Promise.all([getUpdatesPluginApi(), getInstalledPluginApi()])
    .then(([updates, installed]) =>
      updates.plugins.map((updatePlugin: InstalledPlugin) => {
        const updates = getLastUpdates(updatePlugin.updates).map(update =>
          addChangelog(update, updatePlugin.updates)
        );
        const plugin = installed.plugins.find((p: InstalledPlugin) => p.key === updatePlugin.key);
        if (plugin) {
          return {
            ...plugin,
            ...updatePlugin,
            updates
          };
        }
        return { ...updatePlugin, updates };
      })
    )
    .catch(throwGlobalError);
}

export function installPlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/install', data).catch(throwGlobalError);
}

export function uninstallPlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/uninstall', data).catch(throwGlobalError);
}

export function updatePlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/update', data).catch(throwGlobalError);
}

export function cancelPendingPlugins(): Promise<void | Response> {
  return post('/api/plugins/cancel_all').catch(throwGlobalError);
}
