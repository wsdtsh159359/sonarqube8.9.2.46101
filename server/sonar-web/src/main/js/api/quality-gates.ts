/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { BranchParameters } from '../types/branch-like';
import { QualityGateApplicationStatus, QualityGateProjectStatus } from '../types/quality-gates';

export function fetchQualityGates(): Promise<{
  actions: { create: boolean };
  qualitygates: T.QualityGate[];
}> {
  return getJSON('/api/qualitygates/list').catch(throwGlobalError);
}

export function fetchQualityGate(data: { id: number | string }): Promise<T.QualityGate> {
  return getJSON('/api/qualitygates/show', data).catch(throwGlobalError);
}

export function createQualityGate(data: { name: string }): Promise<T.QualityGate> {
  return postJSON('/api/qualitygates/create', data).catch(throwGlobalError);
}

export function deleteQualityGate(data: { id: string }): Promise<void | Response> {
  return post('/api/qualitygates/destroy', data).catch(throwGlobalError);
}

export function renameQualityGate(data: { id: string; name: string }): Promise<void | Response> {
  return post('/api/qualitygates/rename', data).catch(throwGlobalError);
}

export function copyQualityGate(data: { id: string; name: string }): Promise<T.QualityGate> {
  return postJSON('/api/qualitygates/copy', data).catch(throwGlobalError);
}

export function setQualityGateAsDefault(data: { id: string }): Promise<void | Response> {
  return post('/api/qualitygates/set_as_default', data).catch(throwGlobalError);
}

export function createCondition(
  data: {
    gateId: string;
  } & T.Omit<T.Condition, 'id'>
): Promise<T.Condition> {
  return postJSON('/api/qualitygates/create_condition', data).catch(throwGlobalError);
}

export function updateCondition(data: T.Condition): Promise<T.Condition> {
  return postJSON('/api/qualitygates/update_condition', data).catch(throwGlobalError);
}

export function deleteCondition(data: { id: number }): Promise<void> {
  return post('/api/qualitygates/delete_condition', data);
}

export function getGateForProject(data: { project: string }): Promise<T.QualityGate | undefined> {
  return getJSON('/api/qualitygates/get_by_project', data).then(
    ({ qualityGate }) =>
      qualityGate && {
        ...qualityGate,
        isDefault: qualityGate.default
      },
    throwGlobalError
  );
}

export function searchProjects(data: {
  gateName: string;
  page?: number;
  pageSize?: number;
  query?: string;
  selected?: string;
}): Promise<{
  paging: T.Paging;
  results: Array<{ key: string; name: string; selected: boolean }>;
}> {
  return getJSON('/api/qualitygates/search', data).catch(throwGlobalError);
}

export function associateGateWithProject(data: {
  gateId: string;
  projectKey: string;
}): Promise<void | Response> {
  return post('/api/qualitygates/select', data).catch(throwGlobalError);
}

export function dissociateGateWithProject(data: {
  gateId: string;
  projectKey: string;
}): Promise<void | Response> {
  return post('/api/qualitygates/deselect', data).catch(throwGlobalError);
}

export function getApplicationQualityGate(data: {
  application: string;
  branch?: string;
}): Promise<QualityGateApplicationStatus> {
  return getJSON('/api/qualitygates/application_status', data).catch(throwGlobalError);
}

export function getQualityGateProjectStatus(
  data: {
    projectKey?: string;
    projectId?: string;
  } & BranchParameters
): Promise<QualityGateProjectStatus> {
  return getJSON('/api/qualitygates/project_status', data)
    .then(r => r.projectStatus)
    .catch(throwGlobalError);
}
