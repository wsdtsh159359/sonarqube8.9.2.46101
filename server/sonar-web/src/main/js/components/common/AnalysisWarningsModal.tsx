/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { ButtonLink } from 'sonar-ui-common/components/controls/buttons';
import Modal from 'sonar-ui-common/components/controls/Modal';
import WarningIcon from 'sonar-ui-common/components/icons/WarningIcon';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { dismissAnalysisWarning, getTask } from '../../api/ce';
import { sanitizeStringRestricted } from '../../helpers/sanitize';
import { TaskWarning } from '../../types/tasks';
import { withCurrentUser } from '../hoc/withCurrentUser';

interface Props {
  componentKey?: string;
  currentUser: T.CurrentUser;
  onClose: () => void;
  onWarningDismiss?: () => void;
  taskId?: string;
  warnings?: TaskWarning[];
}

interface State {
  loading: boolean;
  dismissedWarning?: string;
  warnings: TaskWarning[];
}

export class AnalysisWarningsModal extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      loading: !props.warnings,
      warnings: props.warnings || []
    };
  }

  componentDidMount() {
    this.mounted = true;
    if (!this.props.warnings && this.props.taskId) {
      this.loadWarnings(this.props.taskId);
    }
  }

  componentDidUpdate(prevProps: Props) {
    const { taskId, warnings } = this.props;
    if (!warnings && taskId && prevProps.taskId !== taskId) {
      this.loadWarnings(taskId);
    } else if (warnings && prevProps.warnings !== warnings) {
      this.setState({ warnings });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleDismissMessage = async (messageKey: string) => {
    const { componentKey } = this.props;

    if (componentKey === undefined) {
      return;
    }

    this.setState({ dismissedWarning: messageKey });

    try {
      await dismissAnalysisWarning(componentKey, messageKey);

      if (this.props.onWarningDismiss) {
        this.props.onWarningDismiss();
      }
    } catch (e) {
      // Noop
    }

    if (this.mounted) {
      this.setState({ dismissedWarning: undefined });
    }
  };

  loadWarnings = async (taskId: string) => {
    this.setState({ loading: true });
    try {
      const { warnings = [] } = await getTask(taskId, ['warnings']);

      if (this.mounted) {
        this.setState({
          loading: false,
          warnings: warnings.map(w => ({ key: w, message: w, dismissable: false }))
        });
      }
    } catch (e) {
      if (this.mounted) {
        this.setState({ loading: false });
      }
    }
  };

  render() {
    const { currentUser } = this.props;
    const { loading, dismissedWarning, warnings } = this.state;

    const header = translate('warnings');

    return (
      <Modal contentLabel={header} onRequestClose={this.props.onClose}>
        <header className="modal-head">
          <h2>{header}</h2>
        </header>

        <div className="modal-body modal-container js-analysis-warnings">
          <DeferredSpinner loading={loading}>
            {warnings.map(({ dismissable, key, message }) => (
              <div className="panel panel-vertical" key={key}>
                <WarningIcon className="pull-left spacer-right" />
                <div className="overflow-hidden markdown">
                  <span
                    // eslint-disable-next-line react/no-danger
                    dangerouslySetInnerHTML={{
                      __html: sanitizeStringRestricted(message.trim().replace(/\n/g, '<br />'))
                    }}
                  />

                  {dismissable && currentUser.isLoggedIn && (
                    <div className="spacer-top display-flex-inline">
                      <ButtonLink
                        className="link-base-color"
                        disabled={Boolean(dismissedWarning)}
                        onClick={() => {
                          this.handleDismissMessage(key);
                        }}>
                        {translate('dismiss_permanently')}
                      </ButtonLink>
                      {dismissedWarning === key && <i className="spinner spacer-left" />}
                    </div>
                  )}
                </div>
              </div>
            ))}
          </DeferredSpinner>
        </div>

        <footer className="modal-foot">
          <ButtonLink className="js-modal-close" onClick={this.props.onClose}>
            {translate('close')}
          </ButtonLink>
        </footer>
      </Modal>
    );
  }
}

export default withCurrentUser(AnalysisWarningsModal);
