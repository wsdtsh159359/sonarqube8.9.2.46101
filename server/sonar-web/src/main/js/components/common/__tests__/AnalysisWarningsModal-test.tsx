/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { dismissAnalysisWarning, getTask } from '../../../api/ce';
import { mockTaskWarning } from '../../../helpers/mocks/tasks';
import { mockCurrentUser, mockEvent } from '../../../helpers/testMocks';
import { AnalysisWarningsModal } from '../AnalysisWarningsModal';

jest.mock('../../../api/ce', () => ({
  dismissAnalysisWarning: jest.fn().mockResolvedValue(null),
  getTask: jest.fn().mockResolvedValue({
    warnings: ['message foo', 'message-bar', 'multiline message\nsecondline\n  third line']
  })
}));

beforeEach(jest.clearAllMocks);

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ warnings: [mockTaskWarning({ dismissable: true })] })).toMatchSnapshot(
    'with dismissable warnings'
  );
  expect(
    shallowRender({
      currentUser: mockCurrentUser({ isLoggedIn: false }),
      warnings: [mockTaskWarning({ dismissable: true })]
    })
  ).toMatchSnapshot('do not show dismissable links for anonymous');
});

it('should not fetch task warnings if it does not have to', () => {
  shallowRender();
  expect(getTask).not.toBeCalled();
});

it('should fetch task warnings if it has to', async () => {
  const wrapper = shallowRender({ taskId: 'abcd1234', warnings: undefined });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(getTask).toBeCalledWith('abcd1234', ['warnings']);
});

it('should correctly handle dismissing warnings', async () => {
  const onWarningDismiss = jest.fn();
  const wrapper = shallowRender({
    componentKey: 'foo',
    onWarningDismiss,
    warnings: [mockTaskWarning({ key: 'bar', dismissable: true })]
  });

  const click = wrapper.find('ButtonLink.link-base-color').props().onClick;

  expect(click).toBeDefined();

  click!(mockEvent());

  await waitAndUpdate(wrapper);

  expect(dismissAnalysisWarning).toBeCalledWith('foo', 'bar');
  expect(onWarningDismiss).toBeCalled();
});

it('should correctly handle updates', async () => {
  const wrapper = shallowRender();

  await waitAndUpdate(wrapper);
  expect(getTask).not.toBeCalled();

  wrapper.setProps({ taskId: '1', warnings: undefined });
  await waitAndUpdate(wrapper);
  expect(getTask).toBeCalled();

  (getTask as jest.Mock).mockClear();
  wrapper.setProps({ taskId: undefined, warnings: [mockTaskWarning()] });
  expect(getTask).not.toBeCalled();
});

function shallowRender(props: Partial<AnalysisWarningsModal['props']> = {}) {
  return shallow<AnalysisWarningsModal>(
    <AnalysisWarningsModal
      currentUser={mockCurrentUser({ isLoggedIn: true })}
      onClose={jest.fn()}
      warnings={[
        mockTaskWarning({ message: 'warning 1' }),
        mockTaskWarning({ message: 'warning 2' })
      ]}
      {...props}
    />
  );
}
