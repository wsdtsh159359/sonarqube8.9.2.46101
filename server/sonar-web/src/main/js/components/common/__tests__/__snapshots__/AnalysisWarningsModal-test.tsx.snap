// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch task warnings if it has to 1`] = `
<Modal
  contentLabel="warnings"
  onRequestClose={[MockFunction]}
>
  <header
    className="modal-head"
  >
    <h2>
      warnings
    </h2>
  </header>
  <div
    className="modal-body modal-container js-analysis-warnings"
  >
    <DeferredSpinner
      loading={false}
    >
      <div
        className="panel panel-vertical"
        key="message foo"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "message foo",
              }
            }
          />
        </div>
      </div>
      <div
        className="panel panel-vertical"
        key="message-bar"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "message-bar",
              }
            }
          />
        </div>
      </div>
      <div
        className="panel panel-vertical"
        key="multiline message
secondline
  third line"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "multiline message<br>secondline<br>  third line",
              }
            }
          />
        </div>
      </div>
    </DeferredSpinner>
  </div>
  <footer
    className="modal-foot"
  >
    <ButtonLink
      className="js-modal-close"
      onClick={[MockFunction]}
    >
      close
    </ButtonLink>
  </footer>
</Modal>
`;

exports[`should render correctly: default 1`] = `
<Modal
  contentLabel="warnings"
  onRequestClose={[MockFunction]}
>
  <header
    className="modal-head"
  >
    <h2>
      warnings
    </h2>
  </header>
  <div
    className="modal-body modal-container js-analysis-warnings"
  >
    <DeferredSpinner
      loading={false}
    >
      <div
        className="panel panel-vertical"
        key="foo"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "warning 1",
              }
            }
          />
        </div>
      </div>
      <div
        className="panel panel-vertical"
        key="foo"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "warning 2",
              }
            }
          />
        </div>
      </div>
    </DeferredSpinner>
  </div>
  <footer
    className="modal-foot"
  >
    <ButtonLink
      className="js-modal-close"
      onClick={[MockFunction]}
    >
      close
    </ButtonLink>
  </footer>
</Modal>
`;

exports[`should render correctly: do not show dismissable links for anonymous 1`] = `
<Modal
  contentLabel="warnings"
  onRequestClose={[MockFunction]}
>
  <header
    className="modal-head"
  >
    <h2>
      warnings
    </h2>
  </header>
  <div
    className="modal-body modal-container js-analysis-warnings"
  >
    <DeferredSpinner
      loading={false}
    >
      <div
        className="panel panel-vertical"
        key="foo"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Lorem ipsum",
              }
            }
          />
        </div>
      </div>
    </DeferredSpinner>
  </div>
  <footer
    className="modal-foot"
  >
    <ButtonLink
      className="js-modal-close"
      onClick={[MockFunction]}
    >
      close
    </ButtonLink>
  </footer>
</Modal>
`;

exports[`should render correctly: with dismissable warnings 1`] = `
<Modal
  contentLabel="warnings"
  onRequestClose={[MockFunction]}
>
  <header
    className="modal-head"
  >
    <h2>
      warnings
    </h2>
  </header>
  <div
    className="modal-body modal-container js-analysis-warnings"
  >
    <DeferredSpinner
      loading={false}
    >
      <div
        className="panel panel-vertical"
        key="foo"
      >
        <WarningIcon
          className="pull-left spacer-right"
        />
        <div
          className="overflow-hidden markdown"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Lorem ipsum",
              }
            }
          />
          <div
            className="spacer-top display-flex-inline"
          >
            <ButtonLink
              className="link-base-color"
              disabled={false}
              onClick={[Function]}
            >
              dismiss_permanently
            </ButtonLink>
          </div>
        </div>
      </div>
    </DeferredSpinner>
  </div>
  <footer
    className="modal-foot"
  >
    <ButtonLink
      className="js-modal-close"
      onClick={[MockFunction]}
    >
      close
    </ButtonLink>
  </footer>
</Modal>
`;
